libs/comparison_rules.py

```diff
@@ -0,0 +1,149 @@
+from office365.runtime.auth.user_credential import UserCredential
+from office365.sharepoint.client_context import ClientContext
+from office365.sharepoint.files.file import File
+import os
+
+FILE_EXTENSIONS = [".xlsx", ".csv", ".xlsb"]
+
+# Define a ruleset for comparison rules
+with ruleset('comparison_rules'):
+    # Define a rule for checking if the row count of two dataframes is equal
+    @when_all(m.rule == 'equal_row_count')
+    def equal_rows(c):
+        """
+        Checks if the row count of the source and target dataframes is equal.
+
+        If the row count is not equal, prints a validation failure message and sets the result to False.
+        If the row count is equal, prints a validation success message and sets the result to True.
+
+        Parameters:
+        c (Context): The context object, which contains the source and target dataframes.
+        """
+        source = pd.DataFrame(c.m.source)
+        target = pd.DataFrame(c.m.target)
+
+        if len(source) != len(target):
+            print(f"Validation failed: Row count does not match (source: {len(source)}, target: {len(target)})")
+            c.s.result = False
+        else:
+            print(f"Validation succeeded: Row count matches (source: {len(source)}, target: {len(target)})")
+            c.s.result = True
+        c.update(c.s)
+
+    # Define a rule for checking if the rows of two dataframes match
+    @when_all(m.rule == 'rows_match')
+    def rows_match(c):
+        """
+        Checks if the rows of the source and target dataframes match.
+
+        If the rows do not match, prints a message and sets the result to False.
+        If the rows match, prints a message and sets the result to True.
+
+        Parameters:
+        c (Context): The context object, which contains the source and target dataframes.
+        """
+        source = pd.DataFrame(c.m.source)
+        target = pd.DataFrame(c.m.target)
+
+        if source.equals(target):
+            print(f"Rows match between both source and target")
+            c.s.result = True
+        else:
+            print(f"Rows do not match between both source and target")
+            c.s.result = False
+        c.update
+
+    # Define a rule for checking if a dataframe has any null values
+    @when_all(m.rule == 'not_null')
+    def not_null(c):
+        """
+        Checks if the target dataframe has any null values.
+
+        If the dataframe has null values, prints a message and returns False.
+        If the dataframe does not have null values, prints a message and returns True.
+
+        Parameters:
+        c (Context): The context object, which contains the target dataframe.
+        """
+        target = pd.DataFrame(c.m.target)
+        if target.notnull().all().all():
+            print(f"No Nulls")
+            c.s.result = True
+        else:
+            print(f"Empty Values")
+            c.s.result = False
+        c.update
+
+    @when_all(m.rule == 'sharepoint_download')
```

Comment:
This should be in its own ruleset and not part of the comparison rule set. Possibly create a new class called preprocessor.py which would contain the ruleset called preprocessor. comparison_rules.py should only deal with dataframes